src/persist/src/nemesis/direct.rs

```diff
@@ -140,26 +172,47 @@ impl Direct {
     }
 
     fn read_output(&mut self, req: ReadOutputReq) -> Result<ReadOutputRes, Error> {
-        let contents = if let Some(output) = self.output_by_stream_name.get_mut(&req.stream) {
-            output.lock()?.clone()
-        } else {
-            vec![]
+        let mut contents = Vec::new();
+        if let Some(output) = self.output_by_stream_name.get_mut(&req.stream) {
+            for e in output.try_iter() {
+                match e {
+                    TimelyCaptureEvent::Progress(x) => {
+                        // TODO: This isn't even a little bit right, but it
+                        // happens to work.
+                        for (ts, ts_diff) in x {
+                            if ts_diff > 0 {
+                                contents.push(ListenEvent::Sealed(ts));
+                            }
+                        }
+                    }
+                    TimelyCaptureEvent::Messages(_, x) => {
+                        contents.push(ListenEvent::Records(x));
+                    }
+                }
+            }
         };
         Ok(ReadOutputRes { contents })
     }
 
     fn seal(&mut self, req: SealReq) -> Result<(), Error> {
-        let (write, _) = self.stream(&req.stream)?;
-        write.seal(req.ts).recv()
+        let (write, _, probe) = self.stream(&req.stream)?;
+        write.seal(req.ts).recv()?;
+
+        // Force the dataflows to make progress, so we don't end up validating
+        // the very uninteresting case of no output.
+        let probe = probe.clone();
+        self.worker.step_while(|| probe.less_than(&req.ts));
```

Comment:
This makes direct_mem waaaaaaaay slower, which is honestly a little surprising to me given that our storage latency is essentially 0 with MemBlob. Something to investigate here for sure.

Reply:
> I'm really curious about the slowdown in direct_mem though. Does it also slow down direct_file or not really?

Thanks for pushing on this. I was being lazy.

- direct_file seems unaffected (within the measurement bounds of my extremely unscientific experiments).
- direct_mem gets a bit slower with the seal fix (3279 -> 2895 in 20s), which makes sense given that we're now generating more writes that actually do work.
- it then gets a bit slower with the addition of the operator (2895 -> 1924 in 20s), which makes sense given that our snapshot replay impl in that operator currently reads it all in at construction time.
- it goes from (1924 -> 1771/1707/383 in 20s) if i add in only the step call in run/stop/seal respectively. this makes sense to me because the seal one is the only one gating by dataflow progress so we're introducing a barrier.
- where it gets really strange is if i put in the step calls in run and stop (but not seal), they are individually fast, but together go from (1924 -> 178 in 20s). the only theory i had about this was that it accidentally starts blocking on operator work (and thus an activator call), but step in timely is implemented as `self.step_or_park(Some(Duration::from_secs(0)))` which is documented in `step_or_park ` to not block if there's no work to do, so that's debunked. shruggie
- then the commit that moves the listener call earlier goes from (119 -> 157 in 20s), confirming that it does get faster

raw dataz

direct_file before this PR:
```
4 runs so far, 0 failures, over 5s
17 runs so far, 0 failures, over 10s
33 runs so far, 0 failures, over 15s
42 runs so far, 0 failures, over 20s
```

direct_file after this PR
```
9 runs so far, 0 failures, over 5s
19 runs so far, 0 failures, over 10s
29 runs so far, 0 failures, over 15s
37 runs so far, 0 failures, over 20s
```

direct_mem before this PR
```
766 runs so far, 0 failures, over 5s
1600 runs so far, 0 failures, over 10s
2419 runs so far, 0 failures, over 15s
3279 runs so far, 0 failures, over 20s
```

direct_mem after this PR
```
37 runs so far, 0 failures, over 5s
76 runs so far, 0 failures, over 10s
117 runs so far, 0 failures, over 15s
157 runs so far, 0 failures, over 20s
```

direct_mem after the seal fix (causing for more successful writes to be generated)
```
687 runs so far, 0 failures, over 5s
1421 runs so far, 0 failures, over 10s
2160 runs so far, 0 failures, over 15s
2895 runs so far, 0 failures, over 20s
```

direct_mem at the end of the second commit
```
26 runs so far, 0 failures, over 5s
58 runs so far, 0 failures, over 10s
89 runs so far, 0 failures, over 15s
119 runs so far, 0 failures, over 20s
```

no step calls at all
```
462 runs so far, 0 failures, over 5s
953 runs so far, 0 failures, over 10s
1444 runs so far, 0 failures, over 15s
1924 runs so far, 0 failures, over 20s
```

step only in seal
```
94 runs so far, 0 failures, over 5s
187 runs so far, 0 failures, over 10s
283 runs so far, 0 failures, over 15s
383 runs so far, 0 failures, over 20s
```

step only in run
```
420 runs so far, 0 failures, over 5s
872 runs so far, 0 failures, over 10s
1315 runs so far, 0 failures, over 15s
1771 runs so far, 0 failures, over 20s
```

step only in stop
```
364 runs so far, 0 failures, over 5s
803 runs so far, 0 failures, over 10s
1249 runs so far, 0 failures, over 15s
1707 runs so far, 0 failures, over 20s
```

step in run and stop
```
41 runs so far, 0 failures, over 5s
90 runs so far, 0 failures, over 10s
132 runs so far, 0 failures, over 15s
178 runs so far, 0 failures, over 20s
```